Source/Core/Core/AchievementManager.cpp

```diff
@@ -262,6 +267,124 @@ bool AchievementManager::IsHardcoreModeActive() const
   return rc_client_is_processing_required(m_client);
 }
 
+bool AchievementManager::VerifyPatchHash(const Common::SHA1::Digest& digest) const
+{
+  File::IOFile list_file;
+  if (!list_file.Open(
+          fmt::format("{}{}{}", File::GetSysDirectory(), DIR_SEP, APPROVED_LIST_FILENAME), "rb"))
+  {
+    ERROR_LOG_FMT(ACHIEVEMENTS, "Failed to open approved patch hash list. Disabling all patches.");
+    return false;
+  }
+  std::vector<u8> buffer(list_file.GetSize());
+  if (!list_file.ReadBytes(buffer.data(), list_file.GetSize()))
+  {
+    ERROR_LOG_FMT(ACHIEVEMENTS, "Failed to read approved patch hash list. Disabling all patches.");
+    return false;
+  }
+  list_file.Close();
+
+  auto context = Common::SHA1::CreateContext();
+  context->Update(buffer);
+  auto list_hash = context->Finish();
+
+  if (list_hash != APPROVED_LIST_HASH)
+  {
+    ERROR_LOG_FMT(ACHIEVEMENTS, "Approved list hash does not match expected hash");
+    return false;
+  }
+
+  std::string hash;
+  for (size_t ix = 0; ix < digest.size(); ix++)
+  {
+    u8 upper = digest[ix] / 16;
+    hash.append(1, upper + ((upper > 9) ? 'A' - 10 : '0'));
+    u8 lower = digest[ix] % 16;
+    hash.append(1, lower + ((lower > 9) ? 'A' - 10 : '0'));
+  }
+
+  auto section_begin_itr =
+      std::search(buffer.begin(), buffer.end(), m_game_ini_id.begin(), m_game_ini_id.end());
+  auto section_end_itr =
+      std::search(section_begin_itr + 1, buffer.end(), m_game_ini_id.begin(), m_game_ini_id.end());
+  auto hash_itr = std::search(section_begin_itr, section_end_itr, hash.begin(), hash.end());
+
+  if (hash_itr == section_end_itr)
+  {
+    ERROR_LOG_FMT(ACHIEVEMENTS, "Ini hash {} from file {} not found in approved list", hash,
+                  m_game_ini_id);
+    return false;
+  }
+
+  return true;
+}
+
+bool AchievementManager::IsPatchApproved(const PatchEngine::Patch& patch) const
+{
+  if (!IsHardcoreModeActive())
+    return true;
+
+  INFO_LOG_FMT(ACHIEVEMENTS, "Verifying patch {}", patch.name);
+  auto context = Common::SHA1::CreateContext();
+  context->Update(Common::BitCastToArray<u8>(patch.entries.size()).data(), 4);
+  for (const auto& entry : patch.entries)
+  {
+    context->Update(Common::BitCastToArray<u8>(entry.type).data(), 1);
```

Comment:
I would also prefer if this didn't have the hardcoded lengths here. Can you do something like

```
template <size_t length>
static void UpdateFromArray(Common::SHA1::Context& context, const std::array<u8, length>& data)
{
  context.Update(data.data(), data.size());
}
```

```
    UpdateFromArray(*context, Common::BitCastToArray<u8>(entry.type));
    UpdateFromArray(*context, Common::BitCastToArray<u8>(entry.address));
    UpdateFromArray(*context, Common::BitCastToArray<u8>(entry.value));
    UpdateFromArray(*context, Common::BitCastToArray<u8>(entry.comparand));
    UpdateFromArray(*context, Common::BitCastToArray<u8>(entry.conditional));
```

(if you're wondering why this is preferable: your code actually has a bug! `entry.type` is 4 bytes long, not 1)

Reply:
Good to know, somehow I was under the impression that enums were one byte.

Reply:
Remove the `AchievementManager::`, the `static` was fine, you just made it a member function for some reason...?